src/html5_parser/node_data/element_data.rs

```diff
@@ -0,0 +1,104 @@
+use std::collections::hash_map::Iter;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ElementAttributes {
+    attributes: HashMap<String, String>,
+}
+
+impl Default for ElementAttributes {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// This is a very thin wrapper around a HashMap.
+/// Most of the methods are the same besides contains/insert.
+/// This "controls" what you're allowed to do with an element's attributes
+/// so there are no unexpected modifications.
+impl ElementAttributes {
+    pub fn new() -> Self {
+        ElementAttributes {
+            attributes: HashMap::new(),
+        }
+    }
+
+    pub fn contains(&self, name: &str) -> bool {
+        self.attributes.contains_key(name)
+    }
+
+    pub fn insert(&mut self, name: &str, value: &str) {
+        self.attributes.insert(name.to_owned(), value.to_owned());
+    }
+
+    pub fn remove(&mut self, name: &str) {
+        self.attributes.remove(name);
+    }
+
+    pub fn get(&self, name: &str) -> Option<&String> {
+        self.attributes.get(name)
+    }
+
+    pub fn get_mut(&mut self, name: &str) -> Option<&mut String> {
+        self.attributes.get_mut(name)
+    }
+
+    pub fn clear(&mut self) {
+        self.attributes.clear();
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.attributes.is_empty()
+    }
+
+    pub fn iter(&self) -> Iter<'_, String, String> {
+        self.attributes.iter()
+    }
+
+    pub fn clone_attributes(&self) -> HashMap<String, String> {
+        self.attributes.clone()
+    }
```

Comment:
An alternative might be to return a reference and allow the caller to opt into cloning:

```suggestion
    pub fn attributes(&self) -> &HashMap<String, String> {
        &self.attributes
    }
```

I noticed a lot of the methods are `pub`.  We might want to default to private or crate-public as much as possible, so that we keep the exposed API surface minimal for later on.  Doing this has the advantage of letting `rust-analyzer` tell you when a function isn't being used.  (It doesn't do this when the function is marked `pub`.)